backend/remote/backend.go

```diff
@@ -387,78 +408,87 @@ func (b *Remote) DeleteWorkspace(name string) error {
+	options := tfe.WorkspaceListOptions{}
 	switch {
-	case name == backend.DefaultStateName:
-		name = b.workspace
-	case b.prefix != "" && !strings.HasPrefix(name, b.prefix):
-		name = b.prefix + name
+	case b.workspace != "":
+		options.Search = tfe.String(b.workspace)
+	case b.prefix != "":
+		options.Search = tfe.String(b.prefix)
 	}
 
-	if !exists {
-		options := tfe.WorkspaceCreateOptions{
-			Name: tfe.String(name),
-		}
+	// Create a slice to contain all the names.
+	var names []string
 
-		// We only set the Terraform Version for the new workspace if this is
-		// a release candidate or a final release.
-		if version.Prerelease == "" || strings.HasPrefix(version.Prerelease, "rc") {
-			options.TerraformVersion = tfe.String(version.String())
+	for {
+		wl, err := b.client.Workspaces.List(context.Background(), b.organization, options)
+		if err != nil {
+			return nil, err
 		}
 
-		_, err = b.client.Workspaces.Create(context.Background(), b.organization, options)
-		if err != nil {
-			return nil, fmt.Errorf("Error creating workspace %s: %v", name, err)
+		for _, w := range wl.Items {
+			if b.workspace != "" && w.Name == b.workspace {
+				names = append(names, backend.DefaultStateName)
+				continue
+			}
+			if b.prefix != "" && strings.HasPrefix(w.Name, b.prefix) {
+				names = append(names, strings.TrimPrefix(w.Name, b.prefix))
+			}
 		}
-	}
 
-	client := &remoteClient{
-		client:       b.client,
-		organization: b.organization,
-		workspace:    name,
+		// Exit the loop when we've seen all pages.
+		if wl.CurrentPage >= wl.TotalPages {
+			break
+		}
 
-		// This is optionally set during Terraform Enterprise runs.
-		runID: os.Getenv("TFE_RUN_ID"),
+		// Update the page number to get the next page.
+		options.PageNumber = wl.NextPage
 	}
 
-	return &remote.State{Client: client}, nil
+	// Sort the result so we have consistent output.
+	sort.StringSlice(names).Sort()
+
+	return names, nil
 }
 
-// Operation implements backend.Enhanced
+// Operation implements backend.Enhanced.
 func (b *Remote) Operation(ctx context.Context, op *backend.Operation) (*backend.RunningOperation, error) {
-	// Configure the remote workspace name.
+	// Get the remote workspace name.
+	workspace := op.Workspace
 	switch {
 	case op.Workspace == backend.DefaultStateName:
-		op.Workspace = b.workspace
+		workspace = b.workspace
 	case b.prefix != "" && !strings.HasPrefix(op.Workspace, b.prefix):
-		op.Workspace = b.prefix + op.Workspace
+		workspace = b.prefix + op.Workspace
 	}
 
+	// Retrieve the workspace for this operation.
+	w, err := b.client.Workspaces.Read(ctx, b.organization, workspace)
+	if err != nil {
+		return nil, generalError("Failed to retrieve workspace", err)
+	}
+
+	// Check if the workspace allows remote operations or that
+	// TF_FORCE_LOCAL_BACKEND is set. In both cases we will use
+	// the local backend to run the operation.
+	if !w.Operations || os.Getenv("TF_FORCE_LOCAL_BACKEND") != "" {
```

Comment:
Could we instead handle this environment variable inside the `Configure` function and set a flag on the backend object from it?

I think it's clearer to handle inputs from all external sources in one place and then have the rest of these functions just access settings from the `b` object, since then it's easier to quickly understand from `Configure` what different inputs will affect the behavior of this backend.

Reply:
Updated